### PR TITLE
MSDF font outline and shadow effects

### PR DESCRIPTION
**Pull request by e-strokov**
Outline and shadow effects for Text Element:

woop

Outline parameters:
    **outline color** - 'rgba' border color, 'black' by default.
    **outline distance** - scalar value for a border width, makes text bold, relative to a font size. Default value is 0 - outline is disabled.

Shadow parameters:
    **shadow color** - 'rgba' color of shadow effect, 'black' by default.
    **shadow offset** - two components vector to offset font UV coordinates to draw a text shadow. Default value is (0, 0) means effect is disabled. Shadow effect includes outline effect extra width.

I confirm I have signed the [Contributor License Agreement](https://docs.google.com/a/playcanvas.com/forms/d/1Ih69zQfJG-QDLIEpHr6CsaAs6fPORNOVnMv5nuo0cjk/viewform).

### CODE REVIEW

File: src/graphics/program-lib/chunks/msdf.frag [hunk 1]
```diff
@@ -33,19 +33,27 @@ uniform float font_sdfIntensity; // intensity is used to boost the value read fr
 uniform float font_pxrange;      // the number of pixels between inside and outside the font in SDF
 uniform float font_textureWidth; // the width of the texture atlas
 
-vec4 applyMsdf(vec4 color) {
-    float font_size = 16.0; // TODO fix this
+uniform vec4 outline_color;
+uniform float outline_distance;
+uniform vec4 shadow_color;
+uniform vec2 shadow_offset;
+//vec2 shadowVector = vec2(0.0062, -0.0046);
```
Comment (willeastcott) [inline src/graphics/program-lib/chunks/msdf.frag]: Delete commented out code.

File: src/graphics/program-lib/chunks/msdf.frag [hunk 2]
```diff
@@ -33,19 +33,27 @@ uniform float font_sdfIntensity; // intensity is used to boost the value read fr
...
     float sigDist = median(tsample.r, tsample.g, tsample.b);
+    float sigDistShdw = median(ssample.r, ssample.g, ssample.b);
 
     #ifdef USE_FWIDTH
         // smoothing depends on size of texture on screen
         vec2 w = fwidth(vUv0);
-        float smoothing = clamp(map(0.0, 2.0 * font_pxrange / font_textureWidth, w.x), 0.0, 0.5);
+        float smoothing = clamp(w.x * font_textureWidth / font_pxrange, 0.0, 0.5);
     #else
+        float font_size = 16.0; // TODO fix this
```
Comment (willeastcott) [inline src/graphics/program-lib/chunks/msdf.frag]: Does this need to be fixed before we merge?

File: src/framework/components/element/text-element.js [hunk 3]
```diff
@@ -1160,6 +1185,169 @@ Object.assign(pc, function () {
...
+                this._outlineColorUniform[3] = this._outlineColor.a;
+                for (var i = 0, len = this._model.meshInstances.length; i < len; i++) {
+                    var mi = this._model.meshInstances[i];
+                    mi.setParameter('outline_color', this._outlineColorUniform);
+                }
+            }
+        }
+    });
+
+    Object.defineProperty(TextElement.prototype, "outlineDistance", {
```
Comment (willeastcott) [inline src/framework/components/element/text-element.js]: Should this be called outlineWidth or outlineThickness instead? Distance seems a little less obvious.

File: src/graphics/program-lib/chunks/msdf.frag [hunk 4]
```diff
@@ -58,16 +66,22 @@ vec4 applyMsdf(vec4 color) {
...
+    sigDistShdw = map(mapMin, mapMax, sigDistShdw + outline_distance);
 
     float center = 0.5;
-
     // calculate smoothing and use to generate opacity
-    // float smoothing = clamp(font_smoothing * (1.0-roy), 0.0, center);
-    float opacity = smoothstep(center-smoothing, center+smoothing, sigDist);
+    float inside = smoothstep(center-smoothing, center+smoothing, sigDistInner);
+    float outline = smoothstep(center-smoothing, center+smoothing, sigDistOutline);
+    float shadow = smoothstep(center-smoothing, center+smoothing, sigDistShdw);
```
Comment (guycalledfrank) [inline src/graphics/program-lib/chunks/msdf.frag]: Might make sense to put outline/shadow only code to a branch or #ifdef

File: src/graphics/program-lib/chunks/msdf.frag (hunk 2 above)
Comment (e-strokov) [reply src/graphics/program-lib/chunks/msdf.frag]: Old code, moved to a fallback branch where it's only being used.

File: src/graphics/program-lib/chunks/msdf.frag (hunk 1 above)
Comment (e-strokov) [reply src/graphics/program-lib/chunks/msdf.frag]: Done.

File: src/framework/components/element/text-element.js (hunk 3 above)
Comment (daredevildave) [reply src/framework/components/element/text-element.js]: Also, what are the units on this? Are they the same as for shadowOffset?

File: src/framework/components/element/text-element.js [hunk 5]
```diff
@@ -1160,6 +1185,169 @@ Object.assign(pc, function () {
         }
     });
 
+    Object.defineProperty(TextElement.prototype, "outlineColor", {
```
Comment (daredevildave) [inline src/framework/components/element/text-element.js]: We need to define these in the element and ensure they are initialized in initializeComponent in ElementComponentSystem.

File: src/framework/components/element/text-element.js (hunk 3 above)
Comment (willeastcott) [reply src/framework/components/element/text-element.js]: More generally, API ref docs for the new properties need to be added. 😄

**Pull request merged**